### PR TITLE
Personal Crafting

### PR DESCRIPTION
**Pull request by DaveTheHeadcrab**
Adds /tg/'s personal crafting system, replaces our current tablecrafting system.
Press the little hammer icon next to your intent selector to open the crafting menu, which will display a list of stuff you can craft by category.

The tribal category is currently unused, but will be filled with *things* soon.

crafting_menu

Items you have the materials to craft will become filled in.

This PR also removes the PDA slot (And puts the pesonal crafting button in its place), which was never properly implemented anyhow. You can put your PDA in your ID slot again, and you can now remove your ID from your PDA by alt clicking it -- Literally nothing has changed functionally except for the removal of an extraneous slot.

Also adds a neat little overlay for when there's an ID in your PDA.
generic_pda_example cap_pda_example engie_pda_example medic_pda_example sci_pda_example sec_pda_example

:cl:DaveTheHeadcrab
rscadd: Replaces tablecrafting with personal crafting, click the hammer icon by your intent selector to use it.
rscdel: Removes the PDA slot, your PDA can now go in your ID slot (Again).
rscadd: You can now alt click to remove an ID from a PDA, for quick interaction while it's in your ID slot.
rscadd: Adds an overlay for an ID while it's inserted into your PDA.
/:cl:

### CODE REVIEW

**Comment by tigercat2000** [comment]
changelog

**Comment by DaveTheHeadcrab** [comment]
I was working on that, calm your tits.

File: code/__DEFINES/clothing.dm [hunk 1]
```diff
@@ -30,8 +30,7 @@
 #define slot_in_backpack 18
 #define slot_legcuffed 19
 #define slot_r_ear 20
-#define slot_wear_pda 21
-#define slot_tie 22
+#define slot_tie 21
 #define slots_amt 22
```
Comment (tigercat2000) [inline code/__DEFINES/clothing.dm]: change this

File: code/modules/crafting/craft.dm [hunk 2]
```diff
@@ -0,0 +1,341 @@
+/datum/personal_crafting
+	var/busy
+	var/viewing_category = 1 //typical powergamer starting on the Weapons tab
+	var/list/categories = list(CAT_WEAPON,CAT_AMMO,CAT_ROBOT,CAT_FOOD,CAT_MISC,CAT_PRIMAL)
+	//var/datum/action/innate/crafting/button // Holdover from when crafting was still an action button. TODO: Remove this before I commit.
```
Comment (tigercat2000) [inline code/modules/crafting/craft.dm]: :eyes:

File: code/modules/crafting/craft.dm [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+	check_tools - takes recipe, a key-type list, and a user and checks if there are enough tools to do the stuff, checks bugs one level deep
+	construct_item - takes a recipe and a user, call all the checking procs, calls do_after, checks all the things again, calls del_reqs, creates result, calls CheckParts of said result with argument being list returned by deel_reqs
+	del_reqs - takes recipe and a user, loops over the recipes reqs var and tries to find everything in the list make by get_environment and delete it/add to parts list, then returns the said list
+*/
+
+
+
+
+/datum/personal_crafting/proc/check_contents(datum/crafting_recipe/R, list/contents)
+	main_loop:
```
Comment (tigercat2000) [inline code/modules/crafting/craft.dm]: jesus why

File: code/modules/crafting/craft.dm [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+	for(var/A in R.chem_catalists)
+		if(contents[A] < R.chem_catalists[A])
+			return 0
+	return 1
+
+/datum/personal_crafting/proc/get_environment(mob/user)
+	. = list()
+	. += user.r_hand
+	. += user.l_hand
+	if(!istype(user.loc, /turf))
```
Comment (tigercat2000) [inline code/modules/crafting/craft.dm]: isturf(user.loc)

File: code/modules/crafting/craft.dm (hunk 3 above)
Comment (DaveTheHeadcrab) [inline code/modules/crafting/craft.dm]: That's not even what the colon does in this context, are you just trained to hate the character at this point?

**Comment by FreeStylaLT** [comment]
No more obscure, guideless tablecrafting? Thank goodness!

**Comment by Spacemanspark** [comment]
Yessss
:+1:

File: code/modules/crafting/craft.dm [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+		return 1
+	var/list/possible_tools = list()
+	for(var/obj/item/I in user.contents)
+		if(istype(I, /obj/item/weapon/storage))
+			for(var/obj/item/SI in I.contents)
+				possible_tools += SI.type
+		possible_tools += I.type
+	possible_tools += contents
+
+	main_loop:
```
Comment (tigercat2000) [inline code/modules/crafting/craft.dm]: whyyy

File: code/modules/crafting/craft.dm [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+	del_reqs return the list of parts resulting object will recieve as argument of CheckParts proc, on the atom level it will add them all to the contents, on all other levels it calls ..() and does whatever is needed afterwards but from contents list already
+*/
+
+/datum/personal_crafting/proc/del_reqs(datum/crafting_recipe/R, mob/user)
+	var/list/surroundings
+	var/list/Deletion = list()
+	. = list()
+	var/data
+	var/amt
+	main_loop:
```
Comment (tigercat2000) [inline code/modules/crafting/craft.dm]: jesus christ

File: code/modules/crafting/craft.dm [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+		Deletion.Cut(Deletion.len)
+		qdel(DL)
+
+/datum/personal_crafting/proc/craft(mob/user)
+	if(user.incapacitated() || user.lying || istype(user.loc, /obj/mecha))
+		return
+	var/list/surroundings = get_surroundings(user)
+	var/dat = "<h3>Crafting menu</h3>"
+	if(busy)
+		dat += "<div class='statusDisplay'>"
```
Comment (tigercat2000) [inline code/modules/crafting/craft.dm]: these should be one line

File: code/modules/crafting/craft.dm (hunk 3 above)
Comment (tigercat2000) [inline code/modules/crafting/craft.dm]: it's not about the colon
it's about the usage of byond labels for continue
that's almost worse than colons

**Comment by Kyep** [comment]
Like the idea of a crafting button/menu.
Don't like the removal of the PDA slot.

**Comment by theColdflame** [comment]
I agree with @Kyep, the menu is good but the slot should stay.

**Comment by DaveTheHeadcrab** [comment]
Again:
*It changes practically nothing. The PDA can be stored in the ID slot. There's no loss of speed when removing ID cards because you can ALT + click. The PDA slot is, plainly, needless clutter and was added by someone who couldn't code for shit as a part of a never-finished and long-dead feature. People will get over losing the slot.*

**Comment by theColdflame** [comment]
removing the slot heavily buffs detomax without any balance changes to compensate, and (I assume) makes ID cards not appear on shiftclick.

**Comment by DaveTheHeadcrab** [comment]
The PDA will list your job just like the ID will when you examine it.

**Comment by Spacemanspark** [comment]
I personally enjoy having the PDA slot.
It might sound silly, but it was just one of those really, really tiny little features that made me fall in love with Paradise when I first joined.

**Comment by theColdflame** [comment]
ID cards not being visible on the majority of crewmembers makes little to no sense from an RP standpoint. The PDA slot is a minor feature, but there's really no reason to remove it.

**Comment by DaveTheHeadcrab** [comment]
There absolutely is.
A mysterious pocket dimension that only your PDA fits in on every person doesn't make any sense either.

**Comment by theColdflame** [comment]
The PDA attaches to the jumpsuit, hence the slot only being usable with a jumpsuit equipped.

**Comment by Spacemanspark** [comment]
Who's to say that jumpsuits couldn't come with a PDA-specific pocket?
IDs (unless we changed the sprite long long ago) actually appear on the chest portion of a jumpsuit. How does it make sense to put a PDA there?

**Comment by DaveTheHeadcrab** [comment]
And that, somehow, doesn't make sense from the ID slot?

**Comment by theColdflame** [comment]
It makes sense with or without the slot- there's no reason to remove it.

**Comment by DaveTheHeadcrab** [comment]
I plainly stated my reasons for removing it, at this point you're ignoring what I'm saying.

**Comment by pipextiny1** [comment]
**Yes** to the table shit. table crafting is cool and only has like 4 things  
**No** to the PDA shit.  why tho? even if it is just a second Id slot that no one looks at what is the point of removing it?

**Comment by DaveTheHeadcrab** [comment]
*DO YOU PEOPLE EVEN READ WHAT I WRITE ANYMORE?!*

**Comment by pipextiny1** [comment]
No

**Comment by Spacemanspark** [comment]
We love you Dave.
But we also love our PDA slot.

**Comment by DaveTheHeadcrab** [comment]
If you aren't gonna put the effort in to read my reasoning why should I put in the effort to write it again?

**Comment by Spacemanspark** [comment]
I did note your reasoning.
But I would still prefer the slot regardless of that reason.
I don't view it as clutter, honestly.

**Comment by tigercat2000** [comment]
@DaveTheHeadcrab you dropped this `\`

**Comment by DaveTheHeadcrab** [comment]
Wasn't talking to you, Spark, I was talking to @pipextiny1 
Who just said they aren't reading what I'm writing.

**Comment by Spacemanspark** [comment]
Ah, alright.

**Comment by pipextiny1** [comment]
The only thing you said was about a mysterious pocket dimension

**Comment by DaveTheHeadcrab** [comment]
> It changes practically nothing. The PDA can be stored in the ID slot. There's no loss of speed when removing ID cards because you can ALT + click to remove the ID now. The PDA slot is, plainly, needless clutter and was added by someone who couldn't code for shit as a part of a never-finished and long-dead feature. There's no reason for it to exist.

**Comment by pipextiny1** [comment]
Issues like?

**Comment by DaveTheHeadcrab** [comment]
It doesn't show up on examine, for instance. Tiny stuff like that because it was *never properly implemented*.

**Comment by theColdflame** [comment]
Oh- I didn't notice you'd edited your previous posts with arguments rather than making new posts, my bad. 

As far as detomax goes, it's already plenty powerful- it remotely stuns someone for upwards of thirty seconds and possibly delimbs them, as well as deafening them for a significant portion of time. Buffing it really isn't needed, though if IDs aren't getting destroyed I guess it's a moot point.

The removal of the slot because it's 'clutter' is a strange choice; fixing the issues with the code behind it would probably be preferable, and regardless of the slot's history. As you just said yourself, 'tiny stuff'- why not just fix it?

Regardless of emoving the slot should probably be its own PR, or at least stated in the title of this one, given how it's not related to the main content of this PR.

**Comment by DaveTheHeadcrab** [comment]
It is, though.
It's removed to make space for the crafting button, you see.

**Comment by tigercat2000** [comment]
you should probably provide a screenshot of the HUD with the button circled or something

**Comment by pipextiny1** [comment]
> It's removed to make space for the crafting button, you see.
We have table crafting why have a button for it?
I have seen the same thing on other stations

**Comment by theColdflame** [comment]
Said button could easily be placed elsewhere on the UI or just made a verb- I doubt the feature will see enough use to be worth that much UI real estate.

**Comment by DaveTheHeadcrab** [comment]
It *removes* tablecrafting, Coldflame.

**Comment by theColdflame** [comment]
I never mentioned tablecrafting?

**Comment by pipextiny1** [comment]
> Do you people even read what I write anymore?!

**Comment by DaveTheHeadcrab** [comment]
Pardon, that was aimed at @pipextiny1 
Who brought up tablecrafting.

**Comment by Spacemanspark** [comment]
I figured that was the reason behind its removal.
I'd just put the button on another part of the HUD, like what Coldflame said.

**Comment by pipextiny1** [comment]
So is there a point not to just add the new things to table crafting?

**Comment by TheDZD** [comment]
@pipextiny1 The point is that this is the replacement to tablecrafting, that doesn't require a table.

**Comment by pipextiny1** [comment]
I know, but why not just add it to the table crafting?

**Comment by DaveTheHeadcrab** [comment]
This doesn't add new recipes. At all.

**Comment by pipextiny1** [comment]
> Do you people even read what I write anymore?!

**Comment by pipextiny1** [comment]
**Why not add them?** Why not add them to table crafting?

**Comment by DaveTheHeadcrab** [comment]
**Why not add what?!**

**Comment by FreeStylaLT** [comment]
Because table crafting is crap, it's obscure, has no guides and is arbitrary. Also, this does not add any new recipes, it only implements the other crafting system.

**Comment by pipextiny1** [comment]
"no guides" It wont need any when it tells you what you need like the crafting system you want to add ( that could easily be added to table crafting)

**Comment by FreeStylaLT** [comment]
Yet hasn't in literal years, iirc. Now that a better, more refined alternative is presented, I would rather take that.

**Comment by DaveTheHeadcrab** [comment]
This system is more versatile on the back end, more immediately apparent to the new players, and easier to use.

There's nothing tablecrafting can do that this system can't.

**Comment by pipextiny1** [comment]
There's nothing tablecrafting can do that this system can't.
?????????????????????????????????????????????????
Then why add a new system that does the same things?
But not update the old one to make it "more versatile on the back end"?

**Comment by DaveTheHeadcrab** [comment]
At this point I'm convinced you're trolling.

**Comment by pipextiny1** [comment]
At this point I'm convinced you don't understand what I'm saying 
Feel free to @pipextiny when ever you find it out

**Comment by FreeStylaLT** [comment]
Nooot exactly how that works, @pipextiny1 . If something has a shoddy backend, to fix it, you usually have to rewrite the entire thing, given that a more robust system already exists in TG, porting it seemed like the better option than trying to iron out, or as I said, rewrite an existing faulty system.

**Comment by Lady-Luck** [comment]
>Like the idea of a crafting button/menu.
Don't like the removal of the PDA slot.

**Comment by tigercat2000** [comment]
>But not update the old one to make it "more versatile on the back end"?

the thing is
this _is_ the update
we stole tablecrafting from -tg-
they upgraded it then replaced it with this
this is a succesor if not direct update to tablecrafting

**Comment by DaveTheHeadcrab** [comment]
I understand very well what you're saying, @pipextiny1 
And it makes no sense.

>There's nothing tablecrafting can do that this system can't.

What this is saying is that, in moving to this system, nothing is lost from the old tablecrafting system -- all functionality is preserved. Which you, evidently, misunderstood.
This *is* making is "more versatile on the back end" in the most straightforward and simple way possible.

**Comment by Fruerlund** [comment]
Looks good, lets now make a list with more tablecraft items ideas so we can /put/ this more to use!

**Comment by Lady-Luck** [comment]
PDAs and IDs are better separate, in my opinion. I know you tried to mitigate it by adding some QoL features like being able to pop the cards out with a modified click, but at the end of the day it's just more intuitive to keep them as separate entities.

**Comment by pipextiny1** [comment]
What I'm saying is that other servers have a better table crafting thing
image
Why not use this?
Or is it because TG  has it?

**Comment by Krausus** [comment]
First off, why is removing the PDA slot even part of this PR? And why isn't it in the PR's title? The only way the existence of the PDA slot can be even *tenuously* linked to this PR's main focus is that you want to put a button where it is. The personal crafting system seems great, but removing PDA slots on a whim in the same PR goes against our contribution guidelines, which mandate atomic PRs.

Aside from that, I still don't really see a point to removing the PDA slot; there doesn't seem to be any outstanding issues with it, aside from it taking up a spot at the bottom of the screen, which is effectively useless anyway, on account of that row of tiles being obscured by a bunch of other icons. But... its existence also doesn't seem necessary. It's just sorta *there.*

However, I'm opposed to its removal unless it's replaced by PDAs that start in the ID slot with the ID already inserted. That would make it functionally identical from the moment players spawn, without having to screw with existing items that currently spawn into the belt slot.

But still, doesn't belong in this PR. Atomicity.

**Comment by DaveTheHeadcrab** [comment]
@pipextiny1 
If I might direct your attention to the OP for a second..
7d820a22-4aab-11e6-82c6-c72112f2a459

**Comment by Lady-Luck** [comment]
>However, I'm opposed to its removal unless it's replaced by PDAs that start in the ID slot with the ID already inserted. That would make it functionally identical from the moment players spawn, without having to screw with existing items that currently spawn into the belt slot.

Pretty sure that's what Dave is doing.

**Comment by pipextiny1** [comment]
At the top mine says table because I'm using a table not the thing you are using

**Comment by DaveTheHeadcrab** [comment]
Right.
Because the only functional difference is one arbitrarily requires a table to use.

**Comment by tigercat2000** [comment]
@Lady-Luck no, he's making them spawn on the belt slot

**Comment by Lady-Luck** [comment]
>no, he's making them spawn on the belt slot

ech

**Comment by DaveTheHeadcrab** [comment]
Ech what? The belt slot is unused at roundstart for all but a very small handful of jobs.

**Comment by FlippingOut** [comment]
As I don't use /tg/'s system, how do you go about using Personal Crafting?
I get you click the button, then the menu pops up, but where do the tools/items come in? Must they be placed within 1 tile of the user, or are they found on the player, and can use items that are in a belt/bag?

Secondly, I don't really care about the PDA slot, but is there a way to put the ID in the PDA on spawn, as I don't like the beltspawn.

Finally, -1 for the new PDA icons. They look really nice, but now Sec can find my hidden ID more often. How can I hide that all-access ID I conned out of the HoP now? Henk.

**Comment by FlippingOut** [comment]
@DaveTheHeadcrab Where does the PDA go on Engineers?

**Comment by DaveTheHeadcrab** [comment]
Belt, their toolbelts are in their bags. I don't mind changing that if people don't like it, that's not my big concern here.

As for personal crafting: All the tools and required materials must be within touch proximity. The craftable recipe will become clickable when that is the case. Then you click it and it works just like tablecrafting. I do not believe items in your belt/bag work for crafting, but if it's in your hand it should.

**Comment by FlippingOut** [comment]
Oh, sounds neat, then.

Now we can finally see people actively use things like improvised shotguns and slugs. It was a bit silly when only older players and those who read the code could make some things and robust everyone else as they didn't know how to combat it.

**Comment by Lady-Luck** [comment]
I'm sorry but the PDA stuff is killing this PR for me. I like that they're separate, Lots of tiny annoyances, like having to stick my ID in there and move it to my ID slot at the start of every round, pop them out of other people's PDAs (and likely having the ID spill out onto the floor), wallets and the cryptographic sequencer toys no longer really being practical.... Not to mention it'll probably be confusing for new players who have enough to learn already.

All for just saving a tiny bit of user interface real estate at the bottom... thanks but no thanks, keep them separate please.

**Comment by FlippingOut** [comment]
@Lady-Luck You can do all that fiddling with your PDA while you walk from the shuttle to your department, or while you're collecting equipment in Engineering/Security/Medical. Popping the IDs out isn't hard at all, as you can Alt-Click them, which means you can strip, alt-click and it's in your hand in the exact same time as before.

Oh, and you can still put your PDA in your bag if you want to use the Crypto ID skin...

**Comment by DaveTheHeadcrab** [comment]
>Having to stick my ID in my PDA at roundstart

Not if you don't intend to use your belt. Which NT reps don't, typically. Even so, takes less than a second in total.

>Pop them out of other people's PDAs.

Alt click means that there's literally no extra work for doing this, even if the other person doesn't know how to remove the ID with alt click. And there's a highly visible message that tells them they can when they insert it.

> (and likely having the ID spill out onto the floor)

Who said anything about that?

> wallets and the cryptographic sequencer toys no longer really being practical.

Wallets are still highly useful storage items that fit in an ID slot, and how does this effect csequencer toys at all? You don't *have* to wear the PDA in the ID slot.

>Not to mention it'll probably be confusing for new players who have enough to learn already.

This is how it is on every other server. They don't appear to be having troubles attracting new players.

**Comment by Lady-Luck** [comment]
>Who said anything about that?

Okay, if I'm the HoP, and I'm holding their PDA in my hands, and I alt click it, is it going to spill out onto the floor? Or am I going to have to switch hands to prevent that from happening?

Bear in mind, as the HoP, I usually have my hands full.

**Comment by DaveTheHeadcrab** [comment]
You could just ALT click it while it's on the table instead of picking it up.

**Comment by Lady-Luck** [comment]
>Not if you don't intend to use your belt. Which NT reps don't, typically. Even so, takes less than a second in total.

NT Rep isn't my main, CMO is. I'm just taking a break from it.

**Comment by FlippingOut** [comment]
If you have your hands full as the HoP, then regardless of this change, you will not be picking that ID up until you have an empty hand.

**Comment by Lady-Luck** [comment]
>This is how it is on every other server. They don't appear to be having troubles attracting new players.

So we have to be like every other server? Keeping them separate is better, in my opinion. For all the things you're adding to mitigate removing the slot in terms of speed and convenient, it's still not as convenient as just having them separate.

**Comment by Lady-Luck** [comment]
>If you have your hands full as the HoP, then regardless of this change, you will not be picking that ID up until you have an empty hand.

You misunderstood, my other hand would likely be occupied.

**Comment by DaveTheHeadcrab** [comment]
Again.
You could just ALT click it while it's on your table.
Same as picking it up.

**Comment by Lady-Luck** [comment]
How about just

keeping them separate

**Comment by FlippingOut** [comment]
Then the alt-click makes it go to the empty hand.

Again, if you had an occupied hand, you would have to swap hands to pick up the ID before this change. This saves you having to swap the hands.

Also, keeping it the same as other servers means people won't be wondering why when they swap servers to Paradise, they have to put an ID in the slot they've otherwise been putting a PDA in for years.

**Comment by Lady-Luck** [comment]
>Also, keeping it the same as other servers means people won't be wondering why when they swap servers to Paradise, they have to put an ID in the slot they've otherwise been putting a PDA in for years.

They don't have to do any wondering, they spawn like that.

**Comment by EpicCharger** [comment]
So with the ID always being in the PDA, will it still show up properly for sec HUDs? And will you be able to still open doors with ease without having to pull your ID out each time?

I do like the idea of having a list of what I can craft on me instead of opening another tab or something. So all up for that.

**Comment by DaveTheHeadcrab** [comment]
Yes, Epic. Functionality is exactly the same. You can just walk into doors. SecHUDs still work as you would expect.

**Comment by Lady-Luck** [comment]
Aaand what if their ID was terminated but their PDA wasn't updated?

**Comment by DaveTheHeadcrab** [comment]
Then that's the HoP's fault.
And they still don't have access.

**Comment by Lady-Luck** [comment]
It probably would be the HoP's fault, yes. At the end of the day though, I'm dealing with someone else's mistake that wouldn't have happened if it was just. separate.

**Comment by TheDZD** [comment]
I honestly see no legitimate reason to keep the slot other than "unique speshul :snowflake: feature" or "it's always been like this, don't change it".

**Comment by pipextiny1** [comment]
@DaveTheHeadcrab Hippie station

**Comment by FlippingOut** [comment]
@Lady-Luck If you'd like to test this, put a terminated ID into a PDA that has been updated with, say, Captain level access. You will not have access, the PDA will just have a title on it.

**Comment by Lady-Luck** [comment]
Why not just put the crafting button where the 'swap' and 'e' buttons are on the hands? You can put it in the upper right and make 'swap' smaller.

**Comment by FlippingOut** [comment]
But then the already small button is replaced with two yet smaller buttons.

**Comment by Lady-Luck** [comment]
What? No, there wouldn't be any replacing, just make the swap button smaller and then put a crafting button there.

**Comment by FlippingOut** [comment]
Yes, making the swap button smaller and putting an equally tiny button next to it.

**Comment by EpicCharger** [comment]
:+1:  Then I'm up for it. I like the idea of a crafting system with you. I don't really see a point in removing the PDA slot, but I also don't really see a point in keeping it either.

**Comment by Lady-Luck** [comment]
>Yes, making the swap button smaller and putting an equally tiny button next to it.

The E button is tiny. The swap button is not. The swap button could be made the size of the 'E' button and the craft button could fill up the rest of the space. That would look good.

**Comment by FlippingOut** [comment]
No, the E button is minuscule, the swap button is tiny. You want to have 3 buttons about the same size in that little crevice above the hand interface.

**Comment by Lady-Luck** [comment]
I'll photoshop it and show you,

**Comment by FreeStylaLT** [comment]
this literally hit over 100 sub-mails in my mailbox and swapped over to a new e-mail.

Jesus Dave just put the PDA slot back, make it a second PR or something, spare us of this madness

**Comment by DaveTheHeadcrab** [comment]
No brakes on this madness train, baby.

**Comment by FreeStylaLT** [comment]
huh that looks pretty neat

**Comment by FlippingOut** [comment]
@Lady-Luck  That looks really neat and wasn't what I pictured from your words. If Dave can implement that, I'd be all for keeping the PDA slot.

**Comment by FlippingOut** [comment]
Actually, upon thinking it over for more than 10 seconds, it should be e swap and c, not e craft and s, as the swap button will get more use, since crafting is a more niche thing in the whole game.

**Comment by Lady-Luck** [comment]
People don't really use the swap button unless they're watching an ERP and they're left handed. ;O

**Comment by theColdflame** [comment]
E/SWAP/C would probably be better, but the idea is very good.

**Comment by Lady-Luck** [comment]
I've got another idea, one moment.

**Comment by razekpraxis** [comment]
An alternative would be to add a third button on the drop/throw area, with a lil hammer

**Comment by Lady-Luck** [comment]
>An alternative would be to add a third button on the drop/throw area, with a lil hammer

That's what I'm working on, but it won't involve adding a new button. Uno momento.

**Comment by FlippingOut** [comment]
>watching an ERP
Literally 2k16, who watches anymore? User involvement is important in game design.

**Comment by FlippingOut** [comment]
Help, I can't format unless it's oldstyle bb.

**Comment by Lady-Luck** [comment]
Rough proof of concept

**Comment by Belario** [comment]
:+1: for the PDA/ID merge, I've always found it odd we have two slots when one can do for both.

**Comment by KasparoVy** [comment]
I think @Lady-Luck's most recent mock up is the best of all the variations I've seen so far. I support this PR as it stands, but I'd prefer it laid out the way @Lady-Luck has it in that most recent post.

**Comment by fludd12** [comment]
oh my god I love you, dave

**Comment by Fox-McCloud** [comment]
:+1: to this.

I don't see the point of keeping the PDA slot either when the way it's laid out here, it has the same functionality currently and was a feature that was added for something that doesn't even exist anymore and doesn't have a point.

Balance wise, not much changes either, as the ID ejects when the PDA is destroyed.

:+1: x :100: to personal crafting. This also means we can kill off `attacky_by` crafting, eventually, too.

**Comment by Spacemanspark** [comment]
Holy shit I'm glad I have E-mail notifications off.
I still want the PDA slot though.

**Comment by IcyV** [comment]
:+1: All for everything here save for the PDA model change when it has an ID. I don't think that will really add much other than entirely remove the PDA ability to hide IDs from people that don't check things. It feels like an unnecessary addition.

I totally support everything else though. I'd love to see how things are without PDA slots again. They don't really have a technical use other than easy-mode bonus slots. Don't want your PDA seen? Pockets. It was a barely fleshed out system that either needs to get finished or go.

**Comment by Spacemanspark** [comment]
Then it should be finished and not chucked out the window.

**Comment by Fox-McCloud** [comment]
>Then it should be finished and not chucked out the window.

It was literally added because of Customs, and that's purely it. Customs is never coming back. **ever**, ergo, there's nothing to "finish" or "flesh out".

**Comment by Spacemanspark** [comment]
I was referring to bug fixes, mostly, although I can only vaguely remember one issue report that had something to do with the PDA slot, and it was fixed IIRC.

**Comment by Fox-McCloud** [comment]
The only changes I'd personally recommend making:

For jobs (not ERT, just station jobs), have their toolbelts (or bartender's case, bandolier) spawn on them and have their PDA equipped to their left pocket.

**Comment by Signmark** [comment]
Now:

- Slot which can only hold PDA. 
- Slot which can only hold IDs.

In this PR:

- Slot which can hold both ID and PDA at same time. Functionally it's same thing as now - you still got PDA and ID, and you can take both in hands quickly.

File: code/modules/crafting/craft.dm (hunk 3 above)
Comment (RemieRichards) [inline code/modules/crafting/craft.dm]: are you high @tigercat2000? it's this or recursion (definitely worse) or a million "Should I stop?" checks+var sets everywhere.

of those 3, this is the best, there is nothing wrong with loop labels.

**Comment by Ssirius** [comment]
I really don't get the hubbub about the PDA slot, nothing really changes with this, you just stuff your ID in your PDA, and put the PDA on the ID slot, and off you go like nothing happened, if you steal someones PDA, you just alt click it and bam the ID pops out and you just toss the PDA away, seems pretty benign to me.

Big fat :+1: on that sweet TG crafting tho.

File: code/modules/crafting/craft.dm (hunk 3 above)
Comment (VampyrBytes) [inline code/modules/crafting/craft.dm]: @RemieRichards 
I'm with @tigercat2000 on this, loop labels are only a small step away from `goto`
And yes, you can do it easily without the need for recursion or a load of checks + var sets everywhere, and without the need for loop labels... just stop trying to do everything in one proc! If you break it down into sub procs, that 'continue main_loop' becomes 'return'

Comment (RemieRichards) [inline code/modules/crafting/craft.dm]: Knew somebody would try to pull that "It's goto!" bullshit, we've had this discussion before at /tg/, it's nowhere near as abusive as ```goto```, and is perfectly fine to use.

And just as a reminder, computers would be nothing without ```goto```, loops and other constructs compile down to them, you can't avoid them, they're only wrong when abused, and it's incredibly difficult to abuse loop labels (I can't even think of a scenario with a loop label that is considered "abuse", overuse isn't even a problem as they're better than temp vars and extraneous ifs that cause branch prediction to fail, in fact, the only time I wouldn't advocate a loop label is when a loop isn't nested, in which case ```continue```/```break``` without a label is the best option)

Also more procs Vamp? really? DM one of the worst things about DM is it's proc call overhead, it's why recursion isn't viable in DM, and it's why it's not a valid approach for it here either.

tl;dr stop.

Comment (VampyrBytes) [inline code/modules/crafting/craft.dm]: It's funny, you knew someone would use the goto argument, I knew you would use the proc call overhead argument, and if this was in something intensive like atmos or lighting, I'd agree that the loop label would be the best method. But for crafting? Really? You've always got to find the right balance between clean code and efficiency... this is just messy code for the sake of over efficiency. How much extra processing do you really think would be added to an entire round by splitting this?

Comment (RemieRichards) [inline code/modules/crafting/craft.dm]:
```
>loop labels
>messy
```
There's no excuse for slower code when there is a faster more readable alternative.
But clearly you just won't see sense, have your cake, but you're certainly not eating it.

Comment (VampyrBytes) [inline code/modules/crafting/craft.dm]: It's not the loop labels themselves that make it messy and less readable, it's the depth of the nested loops and nested if statements, the side effect of which is the need for the loop labels, and the depth, and therefore the need for the labels would be reduced by sub procing

**Comment by Taaketa** [comment]
Remove the PDA slot, import tablecrafting.

**Comment by Lady-Luck** [comment]
I've changed my mind about this PR.

I do like the aesthetic of having the PDA and ID separate, but I am growing increasingly tired of the PDAs never being up to date with someone's ID, because it is kind of a hassle to do, currently.

By unifying them, it should be more consistent. Granted, they might still have to press the 'update the pda' button but the HoP will likely start doing that for the person.

**Comment by Kevak** [comment]
Wait. So this shit removes the bloody PDA slot? Hell no. Put the crafting thing above the run/walk indicator and leave the PDA slot untouched. PDA is _Extremely_ heavily used.

**Comment by Fox-McCloud** [comment]
>Wait. So this shit removes the bloody PDA slot? Hell no. Put the crafting thing above the run/walk indicator and leave the PDA slot untouched. PDA is Extremely heavily used.

Just to make things clear, the PDA slot and ID slot have been merged into a single functionality, it's not just PDA slot removal and no compensation for it.

**Comment by Kevak** [comment]
Since the PR fixing the clusterfuck is already made, I'll just place my critiques for future reference:

Detomax cartrages destroy ID's contained in PDA's. Bad idea.
Taking away an inventory slot. Bad idea.
Removed PDA hiding for a short timespan. Bad idea.

**Comment by Fox-McCloud** [comment]
>Detomax cartrages destroy ID's contained in PDA's. Bad idea.

Wrong. They eject IDs.

>Taking away an inventory slot. Bad idea.

Wrong, when the slots have been merged into a singularity functionality

>Removed PDA hiding for a short timespan. Bad idea.

Debatable when good sec officers will always check your PDA for ID's anyway.

**Comment by Spacemanspark** [comment]
They did state "Short timespan".
